sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -142,13 +194,77 @@ namespace Azure { namespace Core {
 
   public:
     /**
-     * @brief Constructs a new context with no deadline, and no value associated.
+     * @brief Constructs a context with no deadline, and no value associated.
      *
      */
     Context() : m_contextSharedState(std::make_shared<ContextSharedState>()) {}
```

Comment:
I think I remember now - at one point, the root context, which was the application context, did not have a shared state, i.e. it was `nullptr`, and it was not cancelable, but then after `m_contextSharedState(std::make_shared<ContextSharedState>())` was added, the static `ApplicationContext` is vulnerable to being used before initialization, and constexpr is no longer possible (not that it ever was).

Could you please review if `Context::GetDeadline()` implementation still makes sense?


BTW, I know you've thought this through, back and forth, but what if instead of making this change, we delete the copy constructor, make `Context() = default` and make it private, and keep ApplicationContext `const`, even making it constexpr, if it is possible. Then, we say that `ApplicationContext` is not cancelable at all, and it is the same as Go. It is a less of a breaking change, but we'll have to update `context = {}` to `context = Context::ApplicationContext` in the SDK code, but I think it is closer to the original intention that I've heard at one point in the past, and it looks like copy constructor semantics just accidentally slipped in (thanks, default operations) and was not intended to be there in the first place. Yes, it is a breaking change, but so is yours.

Reply:
At one point, I tried making the copy constructor private, and the consequences of that are pretty horrendous. We have a lot of code that depends on the ability to copy context objects around.

And if *we* have a lot of code that copies contexts, I strongly suspect our customers similarly have a lot of code that does the same.